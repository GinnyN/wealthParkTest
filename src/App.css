body {
  background: rgb(181,46,162);
  background: linear-gradient(45deg, rgba(181,46,162,1) 0%, rgba(90,69,159,1) 51%, rgba(0,255,179,1) 100%);
  font-family: 'Sofia Sans', sans-serif;
}
.App{
  background: white;
  margin: 5vh;
  padding: 10px;
  height: 90vh;
  display: grid;
  grid-template-rows: 1fr 8fr 1fr;
  border-radius: 10px;
  -webkit-box-shadow: 3px 4px 10px 14px rgba(255,255,255,0.8);
  -moz-box-shadow: 3px 4px 10px 14px rgba(255,255,255,0.8);
  box-shadow: 3px 4px 10px 14px rgba(255,255,255,0.8);
  overflow: hidden;
}

.container {
  overflow-y: auto;
  overflow-x: none;
}

.spinner {
  height: 40px;
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.item {
  height: 40px;
  padding: 10px;
  display: flex;
  align-items: flex-end;
}
h2 {
  font-family: 'Josefin Slab', serif;
  margin: 0;
}
h2, small {
  padding-left: 10px;
}
a{
  text-decoration: none;
  color: #101010;
}
a:hover {
  color: rgb(181,46,162);
}

.paginator {
  display: flex;
  justify-content: space-between;
}

.paginator button {
  border-radius: 100px;
  height: 50px;
  width: 50px;
  background-color: white;
  border: 2px solid rgb(181,46,162);
}
.paginator button:hover {
  background: linear-gradient(45deg, rgba(181,46,162,1) 0%, rgba(90,69,159,1) 51%, rgba(0,255,179,1) 100%);
  border: 2px solid white;
  color: white;
}
.fadeInRightClass {
  animation-name: fadeInRight;
  animation-duration: 500ms;
}

.fadeInLeftClass {
  animation-name: fadeInLeft;
  animation-duration: 500ms;
}

@keyframes fadeInRight {
  0%{
    opacity: 1;
  }

  45%{
    opacity: 0.1;
    transform: translate3d(-100%, 0, 0);
  }
  50% {
    transform: translate3d(100%, 0, 0);
  }

  100% {
    opacity: 1;
    transform: translate3d(0, 0, 0);
  }
}

@keyframes fadeInLeft {
  0%{
    opacity: 1;
  }

  45%{
    opacity: 0.1;
    transform: translate3d(100%, 0, 0);
  }
  50% {
    transform: translate3d(-100%, 0, 0);
  }

  100% {
    opacity: 1;
    transform: translate3d(0, 0, 0);
  }
}