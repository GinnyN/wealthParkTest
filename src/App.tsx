import { useEffect, useState, useRef } from 'react';
import { hackerNewsService } from './hackerNews.service';
import Item from './item.component';
import './App.css';

const App = () => {
  const itemByPage = 10;
  const amountOfTitles = 100;
  const [itemArray, setItemArray] = useState([]);
  const [page, setPage] = useState(0);
  const [classAnimation, setClassAnimation] = useState('');
  const container = useRef<HTMLDivElement>({} as HTMLDivElement);

  useEffect(() => {
    hackerNewsService.topStories().then((results) => {
      return results.json();
    }).then((json) => setItemArray(json.slice(0, amountOfTitles)));
    container.current?.addEventListener('animationend', () => {
      setClassAnimation('');
    });
  }, [])

  const flipPage = (forward:boolean) => {
    if(forward) {
      setPage(page + itemByPage);
      setClassAnimation('fadeInRightClass');
    } else {
      setPage(page - itemByPage);
      setClassAnimation('fadeInLeftClass');
    }
  }

  return (
    <div className="App">
      <div><h1>HackerNews Top Stories</h1></div> 
        <div className={`container ${classAnimation}`} ref={container}>
          {itemArray.slice(page, page + itemByPage).map((itm, index) => <Item id={itm} key={index+page}></Item> )}
        </div> 
      <div className='paginator'>
        <button className='back' onClick={() => flipPage(false)} disabled={page <= 0}>⇐</button>
        <div>{(page/itemByPage) + 1}</div>
        <button className='forward' onClick={() => flipPage(true)} disabled={page+itemByPage >= 100}>⇒</button>
      </div>
    </div>
  );
}

export default App;
