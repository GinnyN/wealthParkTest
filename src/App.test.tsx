import { render, screen, waitFor, fireEvent, act } from '@testing-library/react';
const renderer = require('react-test-renderer');
import App from './App';
import Item from './item.component';
/*import { hackerNewsService } from './hackerNews.service';

jest.mock("./hackerNews.service", () => ({
  topStories: jest.fn().mockImplementation(() => {
    return Promise.resolve({ json: Promise.resolve([34486699,34487232,34473742,34485871,34485959,34487066,34486824,34480864,34487875])});
  }),
  item: jest.fn().mockImplementation(() => {
    return Promise.resolve({ json: Promise.resolve({"by":"dwardu","descendants":49,"id":34487066,"kids":[34487442,34488446,34488937,34488853,34488324,34487358,34487449,34488246,34489308,34487321,34488407,34488010,34488136,34487500,34487280,34487277,34488179,34487294],"score":95,"time":1674467343,"title":"macOS CLI tool to configure multi-display resolutions and arrangements","type":"story","url":"https://github.com/jakehilborn/displayplacer"})});
  }),
}));*/


describe('Renders First Page', () => {
  act(() => {
    render(<App />);
  })
  const titleElement = screen.getByText('HackerNews Top Stories');
  const backButton = screen.getByText(/⇐/i).closest('button');
  const forwardButton = screen.getByText(/⇒/i).closest('button');

  it("Title h1 in document", () => expect(titleElement).toBeInTheDocument());
  it("Button Back be disabled", () => expect(backButton).toBeDisabled());
  it("Button Forward be Enabled", () => expect(forwardButton).not.toBeDisabled());
});

describe('Snapshot First Page', () => {
  const tree = renderer
    .create(<App />)
    .toJSON();
    waitFor(() => {
      it("First Load Match Snapshot", () => expect(tree).toMatchSnapshot());
    })
});

describe('Calling Top Stories', () => {
  var saveTitle, save2ndTitle: HTMLCollection;
  var appRender: any;
  var items: any[];
  it("Items rendered first time", async () => {
    appRender = await act(() => {
      return render(<App />);
    });
    return waitFor(() => {
      items = appRender.container.getElementsByClassName('item');
      saveTitle = appRender.container.getElementsByClassName('container');
      return expect(items.length).toBe(10);
    
    });
  });
  it("Items changed after button click", async () => {
    appRender = await act(() => {
      return render(<App />);
    });
    return waitFor(() => {
      act(() => {
        fireEvent.click(appRender.container.getElementsByClassName('forward')[0]);
      });
      save2ndTitle = appRender.container.getElementsByClassName('container')
      return expect(save2ndTitle).not.toBe(items)
    });
  });
});

describe('Rendering Item', () => {
  const appRender = render(<Item id={5888} ></Item>);
  waitFor(() => {
    const item = appRender.container.getElementsByClassName('item');
    const heading = appRender.getByRole('heading', {level: 2});
    it("Render 1 item", () => expect(item.length).toBe(1));
    it("Header Exists", () => expect(heading).toBeTruthy());
  });
});

describe('Snapshot Item',  () => {
  const tree = renderer
    .create(<Item id={5888} />)
    .toJSON();
    waitFor(() => {
      it("Snapshot for item 5888 are equal", () => expect(tree).toMatchSnapshot());
    })
});
