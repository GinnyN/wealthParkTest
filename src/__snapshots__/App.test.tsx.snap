// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot First Page First Load Match Snapshot 1`] = `
<div
  className="App"
>
  <div>
    <h1>
      HackerNews Top Stories
    </h1>
  </div>
  <div
    className="container "
  />
  <div
    className="paginator"
  >
    <button
      className="back"
      disabled={true}
      onClick={[Function]}
    >
      ⇐
    </button>
    <div>
      1
    </div>
    <button
      className="forward"
      disabled={false}
      onClick={[Function]}
    >
      ⇒
    </button>
  </div>
</div>
`;

exports[`Snapshot Item 1`] = `
<div
  className="item"
>
  <img
    alt="Loading..."
    className="spinner"
    src="/src/assets/grid.svg"
  />
</div>
`;

exports[`Snapshot Item Snapshot for item 5888 are equal 1`] = `
<div
  className="item"
>
  <img
    alt="Loading..."
    className="spinner"
    src="/src/assets/grid.svg"
  />
</div>
`;
